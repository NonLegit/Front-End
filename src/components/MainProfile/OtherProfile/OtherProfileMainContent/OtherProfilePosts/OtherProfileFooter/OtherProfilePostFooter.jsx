import ChatBubbleOutlineOutlinedIcon from '@mui/icons-material/ChatBubbleOutlineOutlined';
import ShortcutOutlinedIcon from '@mui/icons-material/ShortcutOutlined';
import MoreHorizOutlinedIcon from '@mui/icons-material/MoreHorizOutlined';

import { useEffect, useState } from 'react';
import { ClickAwayListener, Divider } from '@mui/material';
import BookmarkBorderOutlinedIcon from '@mui/icons-material/BookmarkBorderOutlined';
import VisibilityOffOutlinedIcon from '@mui/icons-material/VisibilityOffOutlined';
import UnfoldLessOutlinedIcon from '@mui/icons-material/UnfoldLessOutlined';
import UnfoldMoreOutlinedIcon from '@mui/icons-material/UnfoldMoreOutlined';
import CheckCircleOutlineOutlinedIcon from '@mui/icons-material/CheckCircleOutlineOutlined';
import BlockOutlinedIcon from '@mui/icons-material/BlockOutlined';
import AdminPanelSettingsOutlinedIcon from '@mui/icons-material/AdminPanelSettingsOutlined';
import CancelPresentationOutlinedIcon from '@mui/icons-material/CancelPresentationOutlined';
import FlagOutlinedIcon from '@mui/icons-material/FlagOutlined';
import BookmarksOutlinedIcon from '@mui/icons-material/BookmarksOutlined';
import VisibilityOffIcon from '@mui/icons-material/VisibilityOff';
import { useNavigate } from 'react-router-dom';
import {
  ElementBox, FooterBox, FooterText,
} from './styles';
import OtherProfilePostFooterListResponsive from './OtherProfilePostFooterListResponsive/OtherProfilePostFooterListResponsive';
import OtherProfileArrowList from './OtherProfileArrowList/OtherProfileArrowList';
import { postReactionsServer } from '../../../../profileServer';
import ModeratorList from '../../../../ModeratorList/ModeratorList';

import { useEditPostContext } from '../../../../../../contexts/EditPostContext';

/**
 * Footer of the post that contain all icons
 *
 * @component OtherProfilePostFooter
 * @property {function} handleExpand - pass the event to the parent to show the paragraph of the post
 * @property {state} expand - pass a state to the parent to change the icon
 * @property {object} entity - contains all info of the post
 * @returns {React.Component} OtherProfilePostFooter
 */
function OtherProfilePostFooter(props) {
  const {
    postid, numComments, handleExpand, expand, saved, hidden, submitted, ownerType, owner,
    points, postVoteStatus, isModList, modState, nsfw, spoiler, locked,
    handleLock, handleSpoiler, handleNsfw, handleApprove, handleRemove, handleSpam,
  } = props;
  const [isHidden, setIsHidden] = useState(hidden);
  const [isSaved, setIsSaved] = useState(saved);
  const [showList2, setShowList2] = useState(false);
  const [modList, setModList] = useState(false);
  const [moderatorList, setModeratorList] = useState(false);

  const navigate = useNavigate();
  const { setEditPost, setCommentPost } = useEditPostContext();

  // handle disable the list when click away
  const handleClick2 = () => {
    setShowList2((prev) => !prev);
  };

  const handleClickAway2 = () => {
    setShowList2(false);
  };

  // switch icon to hidden post and vice verse
  const handleClickHide = () => {
    postReactionsServer(postid, hidden ? 'unhide' : 'hide', hidden);
    setIsHidden((prev) => !prev);
  };

  const handleModList = () => {
    setModeratorList((prev) => !prev);
  };

  const handleModListClickAway = () => {
    setModeratorList(false);
  };

  // switch icon to saved post and vice versa
  const handleClickSave = () => {
    postReactionsServer(postid, saved ? 'unsave' : 'save', saved);
    setIsSaved((prev) => !prev);
  };

  const handleShare = () => {
    navigate(`/submit/${postid}`);
  };

  useEffect(() => {
    setIsSaved(saved);
    setIsHidden(hidden);
    setModList(isModList);
  }, [saved, hidden, modList]);

  const publisher = [
    isSaved ? {
      id: 1, text: 'Unsave', icon: <BookmarksOutlinedIcon />, func: handleClickSave,
    } : {
      id: 1, text: 'Save', icon: <BookmarkBorderOutlinedIcon />, func: handleClickSave,
    },
    isHidden ? {
      id: 2, text: 'Unhide', icon: <VisibilityOffIcon />, func: handleClickHide,
    } : {
      id: 2, text: 'Hide', icon: <VisibilityOffOutlinedIcon />, func: handleClickHide,
    },
  ];

  const moderator = [
    {
      id: 1, text: 'Approved', condition: modState === 'approved', icon: <CheckCircleOutlineOutlinedIcon />, func: handleApprove,
    },
    {
      id: 2, text: 'Removed', condition: modState === 'removed', icon: <BlockOutlinedIcon />, func: handleRemove,
    },
    {
      id: 3, text: 'Spam', condition: modState === 'spammed', icon: <CancelPresentationOutlinedIcon />, func: handleSpam,
    },
  ];
  return (
    <FooterBox>
      <OtherProfileArrowList postid={postid} points={points} postVoteStatus={postVoteStatus} />
      <ElementBox>
        {expand ? <UnfoldLessOutlinedIcon sx={{ rotate: '-45deg' }} onClick={() => { handleExpand(); }} />
          : <UnfoldMoreOutlinedIcon sx={{ rotate: '-45deg' }} onClick={() => { handleExpand(); }} />}
      </ElementBox>
      <Divider orientation="vertical" variant="middle" flexItem />
      {/* number of comments and share section */}

      <ElementBox onClick={() => { setCommentPost(true); setEditPost(false); navigate(`/${ownerType === 'Subreddit' ? 'r' : 'user'}/${owner}/comments/${postid}`); }}>
        <ChatBubbleOutlineOutlinedIcon />
        <FooterText variant="caption" responsiveshare={true.toString()}>
          {numComments}
          {' '}
          Comments
        </FooterText>
      </ElementBox>
      <ElementBox onClick={handleShare}>
        <ShortcutOutlinedIcon />
        <FooterText variant="caption" responsiveshare={true.toString()}>Share</FooterText>
      </ElementBox>

      {/* number of comments and share section */}

      {(submitted && modList) && moderator.map((entity) => (
        <ElementBox
          key={entity.id}
          approved={(entity.condition && entity.text === 'Approved')?.toString()}
          spam={(entity.condition && (entity.text === 'Spam' || entity.text === 'Removed'))?.toString()}
          modicons={true.toString()}
          onClick={entity.func}
        >
          {entity.icon}
          <FooterText variant="caption" condition={true.toString()}>{entity.text}</FooterText>
        </ElementBox>
      ))}

      {(submitted && modList) && (
      <>
        <ClickAwayListener onClickAway={handleModListClickAway}>
          <ElementBox>
            <AdminPanelSettingsOutlinedIcon onClick={handleModList} />
            {moderatorList && (
            <ModeratorList
              postid={postid}
              nsfw={nsfw}
              spoiler={spoiler}
              locked={locked}
              handleNsfw={handleNsfw}
              handleSpoiler={handleSpoiler}
              handleLock={handleLock}
            />
            )}
          </ElementBox>
        </ClickAwayListener>
        <ClickAwayListener onClickAway={handleClickAway2}>
          <ElementBox>
            <MoreHorizOutlinedIcon onClick={handleClick2} />
            {showList2 && (
            <OtherProfilePostFooterListResponsive isSaved={isSaved} postid={postid} />
            )}
          </ElementBox>
        </ClickAwayListener>

      </>
      )}

      {(submitted && !modList) && (
      <>
        {publisher.map((entity) => (
          <ElementBox
            key={entity.id}
            onClick={() => { entity.func(); }}
            data-testid={(entity.id === 2) && 'hidden'}
            condition={true.toString()}
          >
            {entity.icon}
            <FooterText variant="caption" condition={true.toString()} data-testid={(entity.id === 2) && 'text-hide'}>{entity.text}</FooterText>
          </ElementBox>
        )) }
        <ElementBox condition={true.toString()}>
          <FlagOutlinedIcon />
          <FooterText variant="caption" condition={true.toString()}>Report</FooterText>
        </ElementBox>

      </>
      )}

      <ClickAwayListener onClickAway={handleClickAway2}>
        <ElementBox show={true.toString()}>
          <MoreHorizOutlinedIcon onClick={handleClick2} />
          {showList2 && (
          <OtherProfilePostFooterListResponsive isSaved={isSaved} postid={postid} />
          )}
        </ElementBox>
      </ClickAwayListener>

    </FooterBox>
  );
}

export default OtherProfilePostFooter;
