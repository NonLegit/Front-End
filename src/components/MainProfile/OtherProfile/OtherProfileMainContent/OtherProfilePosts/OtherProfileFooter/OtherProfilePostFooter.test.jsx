import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import OtherProfilePostFooter from './OtherProfilePostFooter';
import EditPostContextProvider from '../../../../../../contexts/EditPostContext';

test('test snapshot', async () => {
  const tree = renderer.create(
    <Router>
      <EditPostContextProvider>
        <OtherProfilePostFooter />
      </EditPostContextProvider>
    </Router>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
