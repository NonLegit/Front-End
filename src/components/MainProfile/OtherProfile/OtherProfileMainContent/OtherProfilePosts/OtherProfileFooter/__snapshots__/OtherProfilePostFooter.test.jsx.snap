// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test snapshot 1`] = `
<div
  className="MuiBox-root css-146v974"
>
  <div
    className="MuiBox-root css-xv0eud"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1m9p8mk-MuiSvgIcon-root"
      data-testid="ForwardOutlinedIcon"
      focusable="false"
      onClick={[Function]}
      viewBox="0 0 24 24"
    >
      <path
        d="M14 8.83 17.17 12 14 15.17V14H6v-4h8V8.83M12 4v4H4v8h8v4l8-8-8-8z"
      />
    </svg>
    <span
      className="MuiTypography-root MuiTypography-caption css-gj0u27-MuiTypography-root"
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-hshnfv-MuiSvgIcon-root"
      data-testid="ForwardOutlinedIcon"
      focusable="false"
      onClick={[Function]}
      viewBox="0 0 24 24"
    >
      <path
        d="M14 8.83 17.17 12 14 15.17V14H6v-4h8V8.83M12 4v4H4v8h8v4l8-8-8-8z"
      />
    </svg>
  </div>
  <div
    className="MuiBox-root css-orlaaz"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1or9raf-MuiSvgIcon-root"
      data-testid="UnfoldMoreOutlinedIcon"
      focusable="false"
      onClick={[Function]}
      viewBox="0 0 24 24"
    >
      <path
        d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
      />
    </svg>
  </div>
  <hr
    className="MuiDivider-root MuiDivider-middle MuiDivider-vertical MuiDivider-flexItem css-16bghm4-MuiDivider-root"
  />
  <div
    className="MuiBox-root css-orlaaz"
    onClick={[Function]}
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="ChatBubbleOutlineOutlinedIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M20 2H4c-1.1 0-2 .9-2 2v18l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm0 14H6l-2 2V4h16v12z"
      />
    </svg>
    <span
      className="MuiTypography-root MuiTypography-caption css-1052m2m-MuiTypography-root"
      responsiveshare="true"
    >
       
      Comments
    </span>
  </div>
  <div
    className="MuiBox-root css-orlaaz"
    onClick={[Function]}
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="ShortcutOutlinedIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="m15 5-1.41 1.41L15 7.83 17.17 10H8c-2.76 0-5 2.24-5 5v4h2v-4c0-1.65 1.35-3 3-3h9.17L15 14.17l-1.41 1.41L15 17l6-6-6-6z"
      />
    </svg>
    <span
      className="MuiTypography-root MuiTypography-caption css-1052m2m-MuiTypography-root"
      responsiveshare="true"
    >
      Share
    </span>
  </div>
  <div
    className="MuiBox-root css-1mnbp5s"
    onClick={[Function]}
    onTouchEnd={[Function]}
    show="true"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="MoreHorizOutlinedIcon"
      focusable="false"
      onClick={[Function]}
      viewBox="0 0 24 24"
    >
      <path
        d="M6 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm12 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-6 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
      />
    </svg>
  </div>
</div>
`;
